skimage/measure/_find_contours.py

```diff
@@ -111,12 +117,17 @@ def find_contours(array, level,
     if array.ndim != 2:
         raise ValueError('Only 2D arrays are supported.')
     level = float(level)
+    if mask is not None:
+        mask = np.asarray(mask, dtype=np.uint8)
```

Comment:
Ah, silly me, this is the reason why we don't get the buffer error in https://github.com/scikit-image/scikit-image/pull/4150/files#r326102540.

Using asarray here also implies that passing in a mask of dtype('int64') works as well. Do we want to be this forgiving with the input?

Reply:
I think something definitely needs to change here - the current downcast to uint8 doesn't correspond to python truthiness (for example `np.asarray([256], dtype=np.uint8)` results in `[0]`), which is certainly not intuitive to me.

Two suggestions:
- A forgiving API that accepts any mask dtype and treats non-zero as True, zero as False. I think we could implement this cleanly by making the cython function take the mask as `cython.numeric[:, :]`, and casting to uint8 in the python wrapper only if the input dtype is bool.
- Stick with the current approach and check that mask has dtype bool or uint8.

Thoughts?

Reply:
I think `np.asarray(mask, dtype=bool).view(np.uint8)` would work as intended (treating any value != 0 as True).

However after thinking about this a bit more I'd say we don't do users any favors by being more forgiving than the docstring indicates. I think I read something about "In the face of ambiguity..." somewhere. :smile:  So maybe something like this is better:

```python
if mask is not None and not np.can_cast(mask.dtype, bool, casting="safe"):
    raise TypeError("mask is not a binary array")
```

This might be a little annoying if the user has a `mask` of 1s and 0s (`dtype('uint8')`) but then I'd recommend to just call `.view(bool)` before passing it in.

Reply:
@acarapetis Thanks for addressing this. You can even remove the line `mask = np.asarray(mask, dtype=np.uint8)` (and combine the nested if-statements) here and instead pass `mask.view(np.uint8)` to `iterate_and_store` later on. That way we can be sure that no unnecessary copy is performed on `mask`.

Reply:
> You can even remove the line `mask = np.asarray(mask, dtype=np.uint8)` (and combine the nested if-statements) here and instead pass `mask.view(np.uint8)` to `iterate_and_store` later on.

@acarapetis sorry to prod you on this but could you address this part? Calling `np.asarray(mask, dtype=np.uint8)` on a boolean array will make an unnecessary copy. 

```python
>>> a = np.array([False, False], dtype=bool)
>>> b = np.asarray(a, dtype=np.uint8)
>>> c = a.view(np.uint8)
>>> a[0] = True
>>> a, b, c
(array([ True, False]), array([0, 0], dtype=uint8))
```

We can avoid that by using `.view(np.uint8)` instead. 

I'll merge once this is resolved.

Reply:
I've made this change to mask. Is the same problem not true of `array`, for example if the user passes single-precision float? 

Also, regarding "combining the nested if statements", this would require something like `iterate_and_store(..., mask=mask.view(np.uint8) if mask is not None else None)`, so I just stuck with the nested if.

Reply:
> Is the same problem not true of array, for example if the user passes single-precision float?

Assuming I understand you correctly I'd argue that this is a special case due to a "quirk" in NumPy. Each element in a boolean array actually takes up one full byte (same as uint8). Because we made sure that mask is a boolean array we can safely reinterpret that by creating a view showing us `True` as `1` and `False` as `0` which is just how those values are actually stored in memory.

`array` is a different case because `view` behaves differently for numeric dtypes. Creating views between dtypes with a different itemsize only works by changing the number of elements in the array which we clearly don't want here. Furthermore interpreting say an int64 as a float64 will get interesting but in this case not really useful results. E.g. `np.array([3], "i8").view("f8")` will return `array([1.5e-323])`.

> Also, regarding "combining the nested if statements", this would require something like iterate_and_store(..., mask=mask.view(np.uint8) if mask is not None else None), so I just stuck with the nested if.

You're right. I forgot that we need to accommodate `mask` being `None`.